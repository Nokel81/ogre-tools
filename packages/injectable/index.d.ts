/// <reference types="jest" />
export interface DiContainer extends DiContainerForInjection {
  purge: (injectableKey: Injectable<any, any, any>) => void;

  permitSideEffects: (injectableKey: Injectable<any, any, any>) => void;

  override<
    InjectionInstance extends InjectionTokenInstance,
    InjectionTokenInstance,
    InstantiationParam,
  >(
    injectable: Injectable<
      InjectionInstance,
      InjectionTokenInstance,
      InstantiationParam
    >,
    instantiateStub: Instantiate<InjectionInstance, InstantiationParam>,
  ): void;

  unoverride(injectable: Injectable<any, any, any>): void;

  register(...injectables: Injectable<any, any, any>[]): void;

  deregister(...injectables: Injectable<any, any, any>[]): void;
  preventSideEffects: () => void;
}

export type Instantiate<InjectionInstance, InstantiationParam> = {
  (
    di: DiContainerForInjection,
    param: InstantiationParam extends void ? void : InstantiationParam,
  ): InjectionInstance;
};

export interface InjectionToken<InjectionInstance, InstantiationParam> {
  template: InjectionInstance;
  instantiationParameter: InstantiationParam;
  key: Symbol;
}

export interface Injectable<
  InjectionInstance extends InjectionTokenInstance,
  InjectionTokenInstance,
  InstantiationParam,
> {
  id: string;
  causesSideEffects?: boolean;
  injectionToken?: InjectionToken<InjectionTokenInstance, InstantiationParam>;
  instantiate: Instantiate<InjectionInstance, InstantiationParam>;
  lifecycle: ILifecycle<InstantiationParam>;
}

type InjectableLifecycle<InstantiationParam> = InstantiationParam extends void
  ? {
      lifecycle?: ILifecycle<void>;
    }
  : {
      lifecycle: ILifecycle<InstantiationParam>;
    };

type GetInjectableOptions<
  InjectionInstance extends InjectionTokenInstance,
  InjectionTokenInstance,
  InstantiationParam,
> = InjectableLifecycle<InstantiationParam> &
  Omit<
    Injectable<InjectionInstance, InjectionTokenInstance, InstantiationParam>,
    'lifecycle'
  >;

export function getInjectable<
  InjectionInstance extends InjectionTokenInstance,
  InjectionTokenInstance,
  InstantiationParam = void,
>(
  options: GetInjectableOptions<
    InjectionInstance,
    InjectionTokenInstance,
    InstantiationParam
  >,
): Injectable<InjectionInstance, InjectionTokenInstance, InstantiationParam>;

export function getInjectionToken<
  InjectionInstance,
  InstantiationParam = void,
>(options: {
  id: string;
}): InjectionToken<InjectionInstance, InstantiationParam>;

interface Inject {
  <InjectionInstance>(
    key:
      | Injectable<InjectionInstance, unknown, void>
      | InjectionToken<InjectionInstance, void>,
  ): InjectionInstance;

  <InjectionInstance, InstantiationParam>(
    key:
      | Injectable<InjectionInstance, unknown, InstantiationParam>
      | InjectionToken<InjectionInstance, InstantiationParam>,
    param: InstantiationParam,
  ): InjectionInstance;
}

interface InjectMany {
  <InjectionInstance>(
    key:
      | Injectable<InjectionInstance, unknown, void>
      | InjectionToken<InjectionInstance, void>,
  ): InjectionInstance[];

  <InjectionInstance, InstantiationParam>(
    key:
      | Injectable<InjectionInstance, unknown, InstantiationParam>
      | InjectionToken<InjectionInstance, InstantiationParam>,
    param: InstantiationParam,
  ): InjectionInstance[];
}

export interface DiContainerForInjection {
  inject: Inject;
  injectMany: InjectMany;
  register(...injectables: Injectable<any, any, any>[]): void;
  deregister(...injectables: Injectable<any, any, any>[]): void;
}

export interface ILifecycle<InstantiationParam> {
  getInstanceKey: (di: DiContainer, params: InstantiationParam) => any;
}

declare const storedInstanceKey: unique symbol;
declare const nonStoredInstanceKey: unique symbol;

export const lifecycleEnum: {
  singleton: {
    getInstanceKey: (di: DiContainer) => typeof storedInstanceKey;
  };

  keyedSingleton<InstantiationParam>(
    options: ILifecycle<InstantiationParam>,
  ): typeof options;

  transient: {
    getInstanceKey: (di: DiContainer) => typeof nonStoredInstanceKey;
  };
};

export function createContainer(containerId: string): DiContainer;
