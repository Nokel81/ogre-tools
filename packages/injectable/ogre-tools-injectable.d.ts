/// <reference types="jest" />
declare module '@ogre-tools/injectable' {
  type InferFromInjectable<T> = T extends Injectable<
    unknown,
    infer TInstance,
    infer TInstantiationParameter
  >
    ? [TInstance, TInstantiationParameter]
    : never;

  type TentativeTuple<T> = T extends object ? [T] : [undefined?];

  export interface DependencyInjectionContainer {
    inject<TInjectable extends Injectable<unknown, unknown, unknown>>(
      injectableKey: TInjectable,
      ...instantiationParameter: TentativeTuple<
        InferFromInjectable<TInjectable>[1]
      >
    ): InferFromInjectable<TInjectable>[0];

    inject<TInjectionToken extends InjectionToken<unknown, unknown>>(
      injectionToken: TInjectionToken,
      ...instantiationParameter: TentativeTuple<
        TInjectionToken['instantiationParameter']
      >
    ): TInjectionToken['template'];

    purge: (injectableKey: Injectable<any, any, any>) => void;

    runSetups: () => Promise<void>;

    override<TInjectable extends Injectable<unknown, unknown, unknown>>(
      injectable: TInjectable,
      instantiateStub: (
        di: DependencyInjectionContainer,
        ...instantiationParameter: TentativeTuple<
          InferFromInjectable<TInjectable>[1]
        >
      ) => InferFromInjectable<TInjectable>[0],
    ): void;
  }

  export interface ConfigurableDependencyInjectionContainer
    extends DependencyInjectionContainer {
    register(injectable: Injectable<any, any, any>): void;
    preventSideEffects: () => void;
  }

  type InjectionToken<TInstance, TInstantiationParameter> = {
    template: TInstance;
    instantiationParameter: TInstantiationParameter;
    key: Symbol;
  };

  export function getInjectionToken<
    TInstance,
    TInstantiationParameter = void,
  >(): InjectionToken<TInstance, TInstantiationParameter>;

  export interface Injectable<
    TInjectionToken,
    TInstance,
    TInstantiationParameter,
  > {
    id?: string;
    setup?: (di: DependencyInjectionContainer) => void | Promise<void>;
    causesSideEffects?: boolean;
    lifecycle: lifecycleEnum;

    instantiate: (
      di: DependencyInjectionContainer,
      instantiationParameter: TInstantiationParameter,
    ) => TInstance;

    injectionToken?: TInjectionToken;
  }

  type InferFromToken<T> = T extends InjectionToken<
    infer TInstance,
    infer TInstantiationParameter
  >
    ? [TInstance, TInstantiationParameter]
    : never;

  export function getInjectable<
    TInjectionToken extends InjectionToken<TInstance, TInstantiationParameter>,
    TInstance = TInjectionToken extends InjectionToken<any, any>
      ? InferFromToken<TInjectionToken>[0]
      : unknown,
    TInstantiationParameter = TInjectionToken extends InjectionToken<any, any>
      ? InferFromToken<TInjectionToken>[1]
      : unknown,
  >(
    options: Injectable<TInjectionToken, TInstance, TInstantiationParameter>,
  ): Injectable<TInjectionToken, TInstance, TInstantiationParameter>;

  export enum lifecycleEnum {
    singleton,
    transient,
  }

  export function createContainer(
    ...getRequireContexts: any[]
  ): ConfigurableDependencyInjectionContainer;
}
