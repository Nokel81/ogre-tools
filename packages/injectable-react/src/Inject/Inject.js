import { action, observable } from 'mobx';
import { Observer } from 'mobx-react';
import React from 'react';
import constant from 'lodash/fp/constant';

import { isPromise } from '@ogre-tools/fp';

const { Provider: DiContextProvider, Consumer: DiContextConsumer } =
  React.createContext();

export { DiContextProvider };

export default ({
  Component,
  injectableKey = Component,
  getPlaceholder = constant(null),
  ...props
}) => (
  <DiContextConsumer>
    {({ di }) => {
      const maybeAsyncJsx = di.inject(injectableKey, props);

      if (!isPromise(maybeAsyncJsx)) {
        return maybeAsyncJsx;
      }

      const observablePromise = getObservablePromise(maybeAsyncJsx);

      return (
        <Observer>
          {() => {
            const syncJsx = observablePromise.value;

            if (!syncJsx) {
              return getPlaceholder();
            }

            return syncJsx;
          }}
        </Observer>
      );
    }}
  </DiContextConsumer>
);

const getObservablePromise = asyncValue => {
  const observableObject = observable({ value: null }, null, {
    deep: false,
  });

  asyncValue.then(
    action(syncValue => {
      observableObject.value = syncValue;
    }),
  );

  return observableObject;
};
