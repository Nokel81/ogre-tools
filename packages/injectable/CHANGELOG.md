# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [4.1.0](https://github.com/ogre-works/ogre-tools/compare/v4.0.0...v4.1.0) (2022-02-08)


### Features

* Expose keyed singleton as lifecycle in types ([282762e](https://github.com/ogre-works/ogre-tools/commit/282762e3b8e7998fd26474198c10592c67c89ddb))
* introduce keyedSingleton as lifecycle ([174472c](https://github.com/ogre-works/ogre-tools/commit/174472c296c8bdaf2b5ca38469d1f9b3a1963277))
* Permit primitive instantiation parameters ([b10dd82](https://github.com/ogre-works/ogre-tools/commit/b10dd82d7674952129ee5eeb31f50e0c5cb06e6c))



## [4.0.0](https://github.com/ogre-works/ogre-tools/compare/v3.2.1...v4.0.0) (2022-02-07)


### âš  BREAKING CHANGES

* force breaking change to cause major version bump in Lerna

### Bug Fixes

* fix build by using webpack made available to a package by lerna ([5b08e24](https://github.com/ogre-works/ogre-tools/commit/5b08e2472fe06514901546e8a5eb8d8664282a0c))
* force breaking change to cause major version bump in Lerna ([f0c1def](https://github.com/ogre-works/ogre-tools/commit/f0c1defeb85166ccf1907e5d6e9a36cb1cb2314a))


### Reverts

* Revert "Revert "Enhance error for injecting non-registered injectable"" ([8e24a0c](https://github.com/ogre-works/ogre-tools/commit/8e24a0c058758a4c8c50209e9702fd7a2bd792f9))
* Revert "Revert "Implement di.injectMany() as way to inject multiple registered injectables for an injection token"" ([f5d0e01](https://github.com/ogre-works/ogre-tools/commit/f5d0e013d7e079deb5fe1eba5396595850930901))
* Revert "Revert "Permit injection using only injectable or injection token. Replace "id" with "module" for better error messages."" ([fc67c38](https://github.com/ogre-works/ogre-tools/commit/fc67c38aaf4fd4bf81edbcd8dd8bc843864ba0ce))
* Revert "Revert "Remove concept of "injectable viability" for being YAGNI"" ([e4979c6](https://github.com/ogre-works/ogre-tools/commit/e4979c6ef6a002e0f8ae1da248323a2cba672049))
* Revert "Revert "Prevent public access to injection context"" ([4c88b4a](https://github.com/ogre-works/ogre-tools/commit/4c88b4a1c7db1135a2a47c5f97dd976c085bdb42))
