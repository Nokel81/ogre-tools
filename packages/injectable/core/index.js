import getInjectionToken from './src/getInjectionToken/getInjectionToken';

import getInjectable from './src/getInjectable/getInjectable';
import lifecycleEnum from './src/dependency-injection-container/lifecycleEnum';

import createContainer from './src/dependency-injection-container/createContainer';

import isInjectable from './src/getInjectable/isInjectable';
import isInjectionToken from './src/getInjectionToken/isInjectionToken';
import {
  deregistrationCallbackToken,
  injectionDecoratorToken,
  instantiationDecoratorToken,
  registrationCallbackToken,
} from './src/dependency-injection-container/tokens';

export {
  createContainer,
  getInjectable,
  getInjectionToken,
  injectionDecoratorToken,
  instantiationDecoratorToken,
  registrationCallbackToken,
  deregistrationCallbackToken,
  isInjectable,
  isInjectionToken,
  lifecycleEnum,
};
