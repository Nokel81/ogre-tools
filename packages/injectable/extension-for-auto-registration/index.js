import fs from 'fs';
import path from 'path';

import autoRegisterFor from './src/autoRegister';

export const autoRegister = autoRegisterFor({ fs, path });
