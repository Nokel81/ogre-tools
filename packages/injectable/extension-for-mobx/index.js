import {
  computedInjectManyInjectable,
  computedInjectManyWithMetaInjectable,
  registerMobX,
} from './src/computedInjectMany';

export {
  registerMobX,
  computedInjectManyInjectable,
  computedInjectManyWithMetaInjectable,
};
