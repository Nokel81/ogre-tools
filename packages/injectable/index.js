import getInjectionToken from './src/getInjectionToken/getInjectionToken';
import getInjectable from './src/getInjectable/getInjectable';
import lifecycleEnum from './src/dependency-injection-container/lifecycleEnum';
import createContainer, {
  errorMonitorInjectionToken,
} from './src/dependency-injection-container/createContainer';

export {
  getInjectionToken,
  getInjectable,
  lifecycleEnum,
  createContainer,
  errorMonitorInjectionToken,
};
