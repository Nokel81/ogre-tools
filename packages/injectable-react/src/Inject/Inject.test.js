import React from 'react';
import asyncFn from '@async-fn/jest';
import Inject, { DiContextProvider } from './Inject';
import enzyme from 'enzyme';
import { setImmediate as flushMicroTasks } from 'timers';
import { createContainer, lifecycleEnum } from '@ogre-tools/injectable';

const flushPromises = () => new Promise(flushMicroTasks);

const enzymeUpdate = async component => {
  component.setProps();

  await flushPromises();

  component.find('Observer').forEach(Observer => Observer.update());
};

const mountFor =
  di =>
  (node, ...rest) =>
    enzyme.mount(<DiContextProvider value={{ di }}>{node}</DiContextProvider>, {
      ...rest,
    });

describe('Inject', () => {
  let di;
  let mount;

  beforeEach(() => {
    di = createContainer();

    mount = mountFor(di);
  });

  it('given component, when rendering Inject, renders component with injected dependencies', () => {
    const TestComponent = ({ someDependency, ...props }) => (
      <div {...props}>Some content: "{someDependency}"</div>
    );

    di.register({
      id: 'irrelevant',

      lifecycle: lifecycleEnum.transient,

      getDependencies: () => ({
        someDependency: 'some-synchronous-dependency-value',
      }),

      instantiate: (dependencies, props) => (
        <TestComponent {...dependencies} {...props} />
      ),

      aliases: [TestComponent],
    });

    const component = mount(
      <Inject Component={TestComponent} data-some-prop-test />,
    );

    expect(component).toMatchHtmlSnapshot();
  });

  it('given nested Injects, renders', () => {
    const RootTestComponent = ({ someDependency, ...props }) => (
      <div {...props}>
        Some root content: "{someDependency}"
        <Inject
          injectableKey={NestedTestComponent}
          data-some-nested-prop-test
        />
      </div>
    );

    di.register({
      id: 'irrelevant',

      lifecycle: lifecycleEnum.transient,

      getDependencies: () => ({
        someDependency: 'some-root-dependency-value',
      }),

      instantiate: (dependencies, props) => (
        <RootTestComponent {...dependencies} {...props} />
      ),

      aliases: [RootTestComponent],
    });

    const NestedTestComponent = ({ someDependency, ...props }) => (
      <div {...props}>Some nested content: "{someDependency}"</div>
    );

    di.register({
      id: 'irrelevant',

      lifecycle: lifecycleEnum.transient,

      getDependencies: () => ({
        someDependency: 'some-nested-dependency-value',
      }),

      instantiate: (dependencies, props) => (
        <NestedTestComponent {...dependencies} {...props} />
      ),

      aliases: [NestedTestComponent],
    });

    const component = mount(
      <Inject Component={RootTestComponent} data-some-root-prop-test />,
    );

    expect(component).toMatchHtmlSnapshot();
  });

  describe('given no placeholder and nested Injects with an async dependency', () => {
    let component;
    let asyncDependencyMock;

    beforeEach(() => {
      asyncDependencyMock = asyncFn();

      component = getAsyncComponent({
        di,
        asyncDependencyMock,
        placeholder: undefined,
      });
    });

    it('renders with default placeholder', () => {
      expect(component).toMatchHtmlSnapshot();
    });

    it('when the async dependency resolves, renders', async () => {
      await asyncDependencyMock.resolve('some-async-dependency-value');
      await enzymeUpdate(component);

      expect(component).toMatchHtmlSnapshot();
    });
  });

  describe('given placeholder and nested Injects with an async dependency', () => {
    let component;
    let asyncDependencyMock;

    beforeEach(() => {
      asyncDependencyMock = asyncFn();

      component = getAsyncComponent({
        di,
        asyncDependencyMock,
        placeholder: 'some-placeholder',
      });
    });

    it('renders with the placeholder', () => {
      expect(component).toMatchHtmlSnapshot();
    });

    it('when the async dependency resolves, renders', async () => {
      await asyncDependencyMock.resolve('some-async-dependency-value');
      await enzymeUpdate(component);

      expect(component).toMatchHtmlSnapshot();
    });
  });

  it('given non-transient injectable, when rendering, throws', () => {
    const TestComponent = () => 'irrelevant';

    di.register({
      id: 'irrelevant',

      lifecycle: lifecycleEnum.singleton,

      getDependencies: () => ({}),

      instantiate: (dependencies, props) => (
        <TestComponent {...dependencies} {...props} />
      ),

      aliases: [TestComponent],
    });

    jest
      .spyOn(console, 'error')
      .mockImplementationOnce(() => {})
      .mockImplementationOnce(() => {});

    expect(() => {
      mount(<Inject Component={TestComponent} />);
    }).toThrow('Tried to inject non-transient injectable in UI');
  });
});

const getAsyncComponent = ({ di, asyncDependencyMock, placeholder }) => {
  const RootTestComponent = ({ someDependency, ...props }) => (
    <div {...props}>
      Some root content: "{someDependency}"
      <Inject
        {...(placeholder ? { getPlaceholder: () => placeholder } : {})}
        injectableKey={NestedAsyncTestComponent}
        data-some-nested-prop-test
      />
    </div>
  );

  di.register({
    id: 'irrelevant',

    lifecycle: lifecycleEnum.transient,

    getDependencies: () => ({
      someDependency: 'some-root-dependency-value',
    }),

    instantiate: (dependencies, props) => (
      <RootTestComponent {...dependencies} {...props} />
    ),

    aliases: [RootTestComponent],
  });

  const NestedAsyncTestComponent = ({ someDependency, ...props }) => (
    <div {...props}>Some async content: "{someDependency}"</div>
  );

  di.register({
    id: 'irrelevant',

    lifecycle: lifecycleEnum.transient,

    getDependencies: () => ({
      someDependency: asyncDependencyMock(),
    }),

    instantiate: (dependencies, props) => (
      <NestedAsyncTestComponent {...dependencies} {...props} />
    ),

    aliases: [NestedAsyncTestComponent],
  });

  const mount = mountFor(di);

  return mount(
    <Inject Component={RootTestComponent} data-some-root-prop-test />,
  );
};
