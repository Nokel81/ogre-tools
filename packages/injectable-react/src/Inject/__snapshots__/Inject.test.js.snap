// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inject given anonymous function as instantiate, works 1`] = `
<div
  data-dependency-value="some-dependency-value"
  data-instantiation-parameter="instantiation-value"
  data-some-prop-test="true"
>
  Some sync component content
</div>
`;

exports[`Inject given class component with a synchronous dependency, renders 1`] = `
<div
  data-some-prop-test="true"
>
  some-synchronous-dependency-value
</div>
`;

exports[`Inject given class component with an asynchronous dependency, renders 1`] = `
<div
  data-some-prop-test="true"
>
  some-asynchronous-dependency-value
</div>
`;

exports[`Inject given instantiation parameter and one level function component with asynchronous dependencies, works 1`] = `null`;

exports[`Inject given instantiation parameter and one level function component with synchronous dependencies, works 1`] = `
<div
  data-dependency-value="some-synchronous-value"
  data-instantiation-parameter="parameter"
  data-some-prop-test="true"
>
  Some sync component content
</div>
`;

exports[`Inject given no instantiation parameter and one level function component with asynchronous dependencies, works 1`] = `null`;

exports[`Inject given no instantiation parameter and one level function component with synchronous dependencies, works 1`] = `
<div
  data-dependency-value="some-synchronous-value"
  data-some-prop-test="true"
>
  Some sync component content
</div>
`;

exports[`Inject given two level function component with asynchronous dependencies without placeholder does not render component yet 1`] = `null`;

exports[`Inject given two level function component with asynchronous dependencies without placeholder when async thing resolves, renders component 1`] = `
<div
  data-some-prop-test="true"
>
  some-dependency-value
</div>
`;

exports[`Inject given two level function component with synchronous dependencies, works 1`] = `
<div
  data-some-prop-test="true"
>
  Some sync component content
</div>
`;
